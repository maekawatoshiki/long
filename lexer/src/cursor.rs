use sourceloc::SourceLoc;

/// A cursor for a string.
pub(crate) struct Cursor {
    /// The source code the lexer reads.
    source: String,

    /// The current position in `source`.
    pub(crate) pos: usize,

    /// The current source location.
    pub(crate) loc: SourceLoc,
}

impl Cursor {
    /// Creates a new `Cursor`.
    pub fn new(source: impl Into<String>) -> Self {
        Self {
            source: source.into(),
            pos: 0,
            loc: SourceLoc::new(0, 1),
        }
    }

    /// Peeks the character at `self.pos`.
    pub fn peek_char(&self) -> Option<char> {
        self.source.get(self.pos..)?.chars().next()
    }

    /// Peeks the character at `self.pos` and advances `self.pos`.
    pub fn next_char(&mut self) -> Option<char> {
        let c = self.source[self.pos..].chars().next();
        if c == Some('\n') {
            *self.loc.line_mut() += 1;
            *self.loc.column_mut() = 1;
        } else {
            *self.loc.column_mut() += 1;
        }
        self.pos += 1;
        c
    }

    /// Takes a sequence of characters while `preicate()` returns true.
    pub fn take_chars_while<P>(&mut self, mut predicate: P) -> String
    where
        P: FnMut(&char) -> bool,
    {
        let mut buf = String::new();
        loop {
            match self.peek_char() {
                Some(c) if predicate(&c) => {
                    buf.push(self.next_char().unwrap());
                }
                _ => break,
            }
        }
        buf
    }
}
