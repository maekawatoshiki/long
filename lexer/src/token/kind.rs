/// The kind of a token.
#[derive(Debug, Clone)]
pub enum TokenKind {
    Ident(String),
    Keyword(KeywordKind),
    Symbol(SymbolKind),
}

/// The kind of a symbol.
///
/// Currently, some symbols from <https://timsong-cpp.github.io/cppwp/n3337/lex.operators> are
/// not implemented.
#[derive(Debug, Clone)]
pub enum SymbolKind {
    OpeningParen,
    ClosingParen,
    OpeningBrace,
    ClosingBrace,
    OpeningBoxBracket,
    ClosingBoxBracket,
    Comma,
    Semicolon,
    Colon,
    DoubleColon,
    Dot,
    Arrow,
    Inc,
    Dec,
    Add,
    Sub,
    Asterisk,
    Div,
    Mod,
    Not,
    BitwiseNot,
    And,
    Shl,
    Shr,
    Lt,
    Le,
    Gt,
    Ge,
    Eq,
    Ne,
    Xor,
    Or,
    LAnd,
    LOr,
    Question,
    Assign,
    AssignAdd,
    AssignSub,
    AssignMul,
    AssignDiv,
    AssignMod,
    AssignShl,
    AssignShr,
    AssignAnd,
    AssignOr,
    AssignXor,
    AssignLAnd,
    AssignLOr,
    Hash,
    Vararg,
}

/// The kind of a keyword.
#[derive(Debug, Clone)]
pub enum KeywordKind {
    Alignas,
    Continue,
    Friend,
    Register,
    True,
    Alignof,
    Decltype,
    Goto,
    ReinterpretCast,
    Try,
    Asm,
    Default,
    If,
    Return,
    Typedef,
    Auto,
    Delete,
    Inline,
    Short,
    Typeid,
    Bool,
    Do,
    Int,
    Signed,
    Typename,
    Break,
    Double,
    Long,
    Sizeof,
    Union,
    Case,
    DynamicCast,
    Mutable,
    Static,
    Unsigned,
    Catch,
    Else,
    Namespace,
    StatiAssert,
    Using,
    Char,
    Enum,
    New,
    StaticCast,
    Virtual,
    Char16T,
    Explicit,
    Noexcept,
    Struct,
    Void,
    Char32T,
    Export,
    Nullptr,
    Switch,
    Volatile,
    Class,
    Extern,
    Operator,
    Template,
    WcharT,
    Const,
    False,
    Private,
    This,
    While,
    Constexpr,
    Float,
    Protected,
    ThreadLocal,
    ConstCast,
    For,
    Public,
    Throw,
}

impl SymbolKind {
    pub(crate) fn at_least_starts_with(c: char) -> bool {
        matches!(
            c,
            '(' | ')'
                | '{'
                | '}'
                | '['
                | ']'
                | ','
                | ';'
                | ':'
                | '.'
                | '+'
                | '-'
                | '*'
                | '/'
                | '%'
                | '!'
                | '~'
                | '&'
                | '<'
                | '>'
                | '^'
                | '|'
                | '?'
                | '='
                | '#'
        )
    }

    pub(crate) fn from_char(c: char) -> Option<Self> {
        match c {
            '(' => Some(SymbolKind::OpeningParen),
            ')' => Some(SymbolKind::ClosingParen),
            '{' => Some(SymbolKind::OpeningBrace),
            '}' => Some(SymbolKind::ClosingBrace),
            '[' => Some(SymbolKind::OpeningBoxBracket),
            ']' => Some(SymbolKind::ClosingBoxBracket),
            ',' => Some(SymbolKind::Comma),
            ';' => Some(SymbolKind::Semicolon),
            ':' => Some(SymbolKind::Colon),
            '.' => Some(SymbolKind::Dot),
            '+' => Some(SymbolKind::Add),
            '-' => Some(SymbolKind::Sub),
            '*' => Some(SymbolKind::Asterisk),
            '/' => Some(SymbolKind::Div),
            '%' => Some(SymbolKind::Mod),
            '!' => Some(SymbolKind::Not),
            '~' => Some(SymbolKind::BitwiseNot),
            '&' => Some(SymbolKind::And),
            '<' => Some(SymbolKind::Lt),
            '>' => Some(SymbolKind::Gt),
            '^' => Some(SymbolKind::Xor),
            '|' => Some(SymbolKind::Or),
            '?' => Some(SymbolKind::Question),
            '=' => Some(SymbolKind::Assign),
            '#' => Some(SymbolKind::Hash),
            _ => None,
        }
    }

    pub(crate) fn from_two_chars(x: char, y: char) -> Option<SymbolKind> {
        match (x, y) {
            (':', ':') => Some(SymbolKind::DoubleColon),
            ('-', '>') => Some(SymbolKind::Arrow),
            ('+', '+') => Some(SymbolKind::Inc),
            ('-', '-') => Some(SymbolKind::Dec),
            ('<', '<') => Some(SymbolKind::Shl),
            ('>', '>') => Some(SymbolKind::Shr),
            ('<', '=') => Some(SymbolKind::Le),
            ('>', '=') => Some(SymbolKind::Ge),
            ('=', '=') => Some(SymbolKind::Eq),
            ('!', '=') => Some(SymbolKind::Ne),
            ('&', '&') => Some(SymbolKind::LAnd),
            ('|', '|') => Some(SymbolKind::LOr),
            ('+', '=') => Some(SymbolKind::AssignAdd),
            ('-', '=') => Some(SymbolKind::AssignSub),
            ('*', '=') => Some(SymbolKind::AssignMul),
            ('/', '=') => Some(SymbolKind::AssignDiv),
            ('%', '=') => Some(SymbolKind::AssignMod),
            ('&', '=') => Some(SymbolKind::AssignAnd),
            ('|', '=') => Some(SymbolKind::AssignOr),
            ('^', '=') => Some(SymbolKind::AssignXor),
            _ => None,
        }
    }
}

impl KeywordKind {
    /// Returns the kind of the keyword if `kwd` is a keyword.
    pub(crate) fn from_str(kwd: impl AsRef<str>) -> Option<Self> {
        match kwd.as_ref() {
            "alignas" => Some(Self::Alignas),
            "continue" => Some(Self::Continue),
            "friend" => Some(Self::Friend),
            "register" => Some(Self::Register),
            "true" => Some(Self::True),
            "alignof" => Some(Self::Alignof),
            "decltype" => Some(Self::Decltype),
            "goto" => Some(Self::Goto),
            "reinterpret_cast" => Some(Self::ReinterpretCast),
            "try" => Some(Self::Try),
            "asm" => Some(Self::Asm),
            "default" => Some(Self::Default),
            "if" => Some(Self::If),
            "return" => Some(Self::Return),
            "typedef" => Some(Self::Typedef),
            "auto" => Some(Self::Auto),
            "delete" => Some(Self::Delete),
            "inline" => Some(Self::Inline),
            "short" => Some(Self::Short),
            "typeid" => Some(Self::Typeid),
            "bool" => Some(Self::Bool),
            "do" => Some(Self::Do),
            "int" => Some(Self::Int),
            "signed" => Some(Self::Signed),
            "typename" => Some(Self::Typename),
            "break" => Some(Self::Break),
            "double" => Some(Self::Double),
            "long" => Some(Self::Long),
            "sizeof" => Some(Self::Sizeof),
            "union" => Some(Self::Union),
            "case" => Some(Self::Case),
            "dynamic_cast" => Some(Self::DynamicCast),
            "mutable" => Some(Self::Mutable),
            "static" => Some(Self::Static),
            "unsigned" => Some(Self::Unsigned),
            "catch" => Some(Self::Catch),
            "else" => Some(Self::Else),
            "namespace" => Some(Self::Namespace),
            "static_assert" => Some(Self::StatiAssert),
            "using" => Some(Self::Using),
            "char" => Some(Self::Char),
            "enum" => Some(Self::Enum),
            "new" => Some(Self::New),
            "static_cast" => Some(Self::StaticCast),
            "virtual" => Some(Self::Virtual),
            "char16_t" => Some(Self::Char16T),
            "explicit" => Some(Self::Explicit),
            "noexcept" => Some(Self::Noexcept),
            "struct" => Some(Self::Struct),
            "void" => Some(Self::Void),
            "char32_t" => Some(Self::Char32T),
            "export" => Some(Self::Export),
            "nullptr" => Some(Self::Nullptr),
            "switch" => Some(Self::Switch),
            "volatile" => Some(Self::Volatile),
            "class" => Some(Self::Class),
            "extern" => Some(Self::Extern),
            "operator" => Some(Self::Operator),
            "template" => Some(Self::Template),
            "wchar_t" => Some(Self::WcharT),
            "const" => Some(Self::Const),
            "false" => Some(Self::False),
            "private" => Some(Self::Private),
            "this" => Some(Self::This),
            "while" => Some(Self::While),
            "constexpr" => Some(Self::Constexpr),
            "float" => Some(Self::Float),
            "protected" => Some(Self::Protected),
            "thread_local" => Some(Self::ThreadLocal),
            "const_cast" => Some(Self::ConstCast),
            "for" => Some(Self::For),
            "public" => Some(Self::Public),
            "throw" => Some(Self::Throw),
            _ => None,
        }
    }
}
