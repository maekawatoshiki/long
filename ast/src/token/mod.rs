pub mod kind;

use std::collections::HashSet;

use kind::TokenKind;
use long_sourceloc::SourceLoc;

#[derive(Debug, Clone)]
pub struct Token {
    /// The kind of the token.
    kind: TokenKind,

    /// `Some(...)` if any leading spaces occurs before the token.
    /// Line terminator characters are not leading spaces.
    leading_space: bool,

    /// The hideset of the token.
    // TODO: We should create `PPToken` (with a hideset) to simplify code.
    hideset: HashSet<String>,

    /// The source location of the token.
    loc: SourceLoc,
}

impl Token {
    /// Creates a new `Token`.
    #[inline]
    pub fn new(kind: TokenKind, loc: SourceLoc) -> Self {
        Self {
            kind,
            leading_space: false,
            hideset: HashSet::new(),
            loc,
        }
    }

    /// Sets the kind of the token and returns the token.
    #[inline]
    pub fn with_kind(mut self, kind: TokenKind) -> Self {
        self.kind = kind;
        self
    }

    /// Sets a leading space flag and returns the token.
    #[inline]
    pub fn set_leading_space(mut self, leading_space: bool) -> Self {
        self.leading_space = leading_space;
        self
    }

    /// Sets the hideset and returns the token.
    #[inline]
    pub fn with_hideset(mut self, hideset: HashSet<String>) -> Self {
        self.hideset = hideset;
        self
    }

    /// Modifies the hideset by `f` and returns the token.
    #[inline]
    pub fn with_hideset_modified<F>(mut self, f: F) -> Self
    where
        F: FnOnce(&mut HashSet<String>),
    {
        f(&mut self.hideset);
        self
    }

    /// Sets the source location and returns the token.
    #[inline]
    pub fn with_loc(mut self, loc: SourceLoc) -> Self {
        self.loc = loc;
        self
    }

    /// Returns the kind of the token.
    #[inline]
    pub fn kind(&self) -> &TokenKind {
        &self.kind
    }

    /// Returns the source location of the token.
    #[inline]
    pub fn loc(&self) -> &SourceLoc {
        &self.loc
    }

    /// Returns the leading space of the token.
    #[inline]
    pub fn leading_space(&self) -> bool {
        self.leading_space
    }

    /// Returns the hideset of the token.
    #[inline]
    pub fn hideset(&self) -> &HashSet<String> {
        &self.hideset
    }

    /// Converts the token to a string.
    #[inline]
    pub fn to_string(&self) -> String {
        self.kind.to_string()
    }
}

/// Converts the token stream to a string.
pub fn stringify(tokens: &[Token], ignore_leading_space: bool) -> String {
    tokens
        .iter()
        .fold("".to_string(), |acc, tok| {
            format!(
                "{}{}{}",
                acc,
                if !ignore_leading_space && tok.leading_space {
                    " "
                } else {
                    ""
                },
                tok.to_string()
            )
        })
        .trim_start() // Remove a leading space.
        .to_string()
}
