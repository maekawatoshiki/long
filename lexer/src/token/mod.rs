pub mod kind;

use kind::TokenKind;
use long_sourceloc::SourceLoc;

#[derive(Debug, Clone)]
pub struct Token {
    /// The kind of the token.
    kind: TokenKind,

    /// The source location of the token.
    loc: SourceLoc,
}

impl Token {
    /// Creates a new `Token`.
    pub fn new(kind: TokenKind, loc: SourceLoc) -> Self {
        Self { kind, loc }
    }

    /// Returns the kind of the token.
    pub fn kind(&self) -> &TokenKind {
        &self.kind
    }

    /// Returns the source location of the token.
    pub fn loc(&self) -> &SourceLoc {
        &self.loc
    }
}
