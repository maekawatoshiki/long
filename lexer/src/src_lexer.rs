use crate::cursor::Cursor;
use crate::token::kind::{FloatKind, IntKind, KeywordKind, SymbolKind, TokenKind};
use crate::token::Token;
use anyhow::Result;
use std::fmt;
use std::fs::read_to_string;
use std::path::PathBuf;

/// A lexical analyzer for either a single source file or a string.
pub(crate) struct SourceLexer {
    /// The file path of the source file the lexer originally reads.
    /// Set to `None` if the lexer does not read a file but a string.
    #[allow(dead_code)]
    filepath: Option<PathBuf>,

    /// The cursor of the lexer.
    cursor: Cursor,
}

#[derive(Debug)]
pub(crate) enum Error {
    Include(PathBuf),
}

impl SourceLexer {
    /// Creates a new `SourceLexer`.
    pub fn new(source: impl Into<String>) -> Self {
        Self {
            filepath: None,
            cursor: Cursor::new(source.into()),
        }
    }

    /// Creates a new `SourceLexer` for a file.
    pub fn new_from_file<P>(filepath: P) -> Result<Self>
    where
        P: Into<PathBuf> + Clone,
    {
        Ok(Self {
            cursor: Cursor::new(read_to_string(filepath.clone().into())?),
            filepath: Some(filepath.into()),
        })
    }

    /// Reads a token.
    pub fn next(&mut self) -> Result<Option<Token>> {
        match self.cursor.peek_char() {
            Some(c) if c.is_ascii_alphabetic() || c == '_' => Ok(Some(self.read_identifier())),
            Some(c) if c == ' ' || c == '\t' || c == '\n' => {
                let leading_space = !self.read_whitespaces().ends_with('\n');
                self.next()
                    .map(|t| t.map(|t| t.set_leading_space(leading_space)))
            }
            Some(c)
                if c.is_ascii_digit()
                    || (c == '.'
                        && self
                            .cursor
                            .peek_char2()
                            .map_or(false, |c| c.is_ascii_digit())) =>
            {
                Ok(Some(self.read_number()))
            }
            Some(c) if SymbolKind::at_least_starts_with(c) => Ok(Some(self.read_symbol())),
            None => return Ok(None),
            _ => todo!(),
        }
    }

    /// Reads an identifier. Assumes the result of `self.cursor.peek_char()` is alphabetic.
    fn read_identifier(&mut self) -> Token {
        let loc = self.cursor.loc;
        let ident = self
            .cursor
            .take_chars_while(|&c| c.is_ascii_alphanumeric() || c == '_');
        KeywordKind::from_str(ident.as_str()).map_or_else(
            || Token::new(TokenKind::Ident(ident), loc),
            |kind| Token::new(TokenKind::Keyword(kind), loc),
        )
    }

    /// Reads a symbol.
    fn read_symbol(&mut self) -> Token {
        let loc = self.cursor.loc;
        let x = self.cursor.peek_char();
        let y = self.cursor.peek_char2();
        match (x, y) {
            (Some('.'), Some('.')) if self.cursor.peek_char3() == Some('.') => {
                self.cursor.next_char();
                self.cursor.next_char();
                self.cursor.next_char();
                Token::new(TokenKind::Symbol(SymbolKind::Vararg), loc)
            }
            (Some('<'), Some('<'))
            | (Some('>'), Some('>'))
            | (Some('&'), Some('&'))
            | (Some('|'), Some('|'))
                if self.cursor.peek_char3() == Some('=') =>
            {
                self.cursor.next_char();
                self.cursor.next_char();
                self.cursor.next_char();
                Token::new(
                    TokenKind::Symbol(match x {
                        Some('<') => SymbolKind::AssignShl,
                        Some('>') => SymbolKind::AssignShr,
                        Some('&') => SymbolKind::AssignLAnd,
                        Some('|') => SymbolKind::AssignLOr,
                        _ => unreachable!(),
                    }),
                    loc,
                )
            }
            (Some(x), Some(y)) => match SymbolKind::from_two_chars(x, y) {
                Some(kind) => {
                    self.cursor.next_char();
                    self.cursor.next_char();
                    Token::new(TokenKind::Symbol(kind), loc)
                }
                None => self.read_symbol1(),
            },
            _ => self.read_symbol1(),
        }
    }

    /// Reads a single-character symbol.
    fn read_symbol1(&mut self) -> Token {
        let loc = self.cursor.loc;
        let c = self.cursor.next_char().unwrap();
        Token::new(
            TokenKind::Symbol(SymbolKind::from_char(c).expect("unreachable")),
            loc,
        )
    }

    /// Reads a number literal, including integer literals and floating-point literals.
    fn read_number(&mut self) -> Token {
        let loc = self.cursor.loc;
        let mut is_float = false;
        let mut last = '\0';
        let lit = self.cursor.take_chars_while(|&c| {
            let is_f = "eEpP".contains(last) && "+-".contains(c);
            is_float = is_float || c == '.' || is_f;
            last = c;
            c.is_alphanumeric() || c == '.' || is_f
        });
        if is_float {
            // TODO: Support suffix.
            let f: f64 = lit.parse().unwrap();
            Token::new(TokenKind::Float(FloatKind::Double(f)), loc)
        } else {
            // TODO: Support suffix.
            // TODO: Support 64bit int.
            // TODO: Support hex and oct literals.
            let i: i32 = lit.parse().unwrap();
            Token::new(TokenKind::Int(IntKind::Int(i)), loc)
        }
    }

    /// Reads whitespaces (i.e. ' ' '\t' '\n').
    fn read_whitespaces(&mut self) -> String {
        self.cursor
            .take_chars_while(|&c| c == ' ' || c == '\t' || c == '\n')
    }
}

impl std::error::Error for Error {}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}

#[test]
fn test() {
    let _ = SourceLexer::new("int main() {}");
}

#[test]
#[should_panic]
fn cannot_open_file() {
    let _ = SourceLexer::new_from_file("").unwrap();
}

#[test]
fn just_read_one_token() {
    let mut l = SourceLexer::new("int main() {}");
    assert!(matches!(
        l.next().unwrap().unwrap().kind(),
        TokenKind::Keyword(KeywordKind::Int)
    ));
    assert!(l.cursor.pos == 3);
    let mut l = SourceLexer::new("__num");
    assert!(matches!(l.next().unwrap().unwrap().kind(), TokenKind::Ident(i) if i == "__num"));
    assert!(l.cursor.pos == 5)
}

#[test]
fn read_tokens() {
    let mut l = SourceLexer::new("int main\nlong short");
    insta::assert_debug_snapshot!(read_all_tokens(&mut l));
}

#[test]
fn read_tokens2() {
    let mut l = SourceLexer::new("int main()");
    insta::assert_debug_snapshot!(read_all_tokens(&mut l));
}

#[test]
fn read_tokens3() {
    let mut l = SourceLexer::new("int main() {\n return;\n}");
    insta::assert_debug_snapshot!(read_all_tokens(&mut l));
}

#[test]
fn read_symbols() {
    let mut l = SourceLexer::new(
        "(){}[],;:.+-*/%!~&<>^|?=#++--::-><<>><=>===!=&&||+=-=*=/=%=&=^=|=<<=>>=&&=||=sizeof",
    );
    insta::assert_debug_snapshot!(read_all_tokens(&mut l));
}

#[test]
fn read_ints() {
    let mut l = SourceLexer::new(".123 1e+10 34567890 123");
    insta::assert_debug_snapshot!(read_all_tokens(&mut l));
}

#[cfg(test)]
fn read_all_tokens(l: &mut SourceLexer) -> Vec<Token> {
    let mut tokens = vec![];
    while let Some(tok) = l.next().unwrap() {
        tokens.push(tok)
    }
    tokens
}
