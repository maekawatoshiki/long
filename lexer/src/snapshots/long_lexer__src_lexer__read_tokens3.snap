---
source: lexer/src/src_lexer.rs
expression: read_all_tokens(&mut l)

---
[
    Token {
        kind: Keyword(
            Int,
        ),
        leading_space: false,
        loc: 1:1,
    },
    Token {
        kind: Ident(
            "main",
        ),
        leading_space: true,
        loc: 1:5,
    },
    Token {
        kind: Symbol(
            OpeningParen,
        ),
        leading_space: false,
        loc: 1:9,
    },
    Token {
        kind: Symbol(
            ClosingParen,
        ),
        leading_space: false,
        loc: 1:10,
    },
    Token {
        kind: Symbol(
            OpeningBrace,
        ),
        leading_space: true,
        loc: 1:12,
    },
    Token {
        kind: Keyword(
            Return,
        ),
        leading_space: true,
        loc: 2:2,
    },
    Token {
        kind: Symbol(
            Semicolon,
        ),
        leading_space: false,
        loc: 2:8,
    },
    Token {
        kind: Symbol(
            ClosingBrace,
        ),
        leading_space: false,
        loc: 3:1,
    },
]
